### PR TITLE
Fixed #7374 -- Consolidate usage of Excel libraries

### LINKED ISSUE #7374: Reduce the amount of Excel file read/writing libraries
Wagtail has a fairly large number of (in)direct dependencies. By installing Wagtail in an empty virtualenv I ended up with 32 unique Python packages. Of those 32 packages 4 deal with reading/writing Excel files.

1. `XlsxWriter` (referenced directly in Wagtail's `setup.py`)
    _XlsxWriter is a Python module for writing files in the Excel 2007+ XLSX file format_
2. `openpyxl` (referenced directly in Wagtail's `setup.py`, and pulled in by the `tablib[xlsx]` dependency)
    _A Python library to read/write Excel 2010 xlsx/xlsm files_
3. `xlrd` (pulled in by the `tablib[xls]` dependency)
    _xlrd is a library for reading data and formatting information from Excel files in the historical .xls format._
4. `xlwt` (pulled in by the `tablib[xls]` dependency)
    _This is a library for developers to use to generate spreadsheet files compatible with Microsoft Excel versions 95 to 2003._

I personally don't associate Wagtail with Excel, so I was wondering where those dependencies came from.

## XlsxWriter

The only usage I could find in the Wagtail codebase is the (undocumented) `wagtail.admin.views.SpreadsheetExportMixin`. This mixin is used by the Wagtail forms and modeladmin contrib apps.

## openpyxl / xlrd

These two are used by the redirects contrib app to support the import (and export?) of redirects. In addition `openpyxl` is used to test the output generated by `SpreadsheetExportMixin` for the forms and modeladmin contrib apps.

## xlwt

I could not find any usages by Wagtail, the only reason it is installed is because of the `tablib[xls]` dependency.

## Suggestion

1. There is significant overlap between the `XlsxWriter` and `openpyxl` packages. Their APIs are also very similar, the only difference is the minimal supported Excel version (2007 v.s. 2010). Wagtail already uses `openpyxl` to verify the output created by `XlsxWriter`. Would it be acceptable to rewrite the `SpreadsheetExportMixin` to use `openpyxl` instead of `XlsxWriter`, and drop the `XlsxWriter` dependency?

2. Drop support for the import (and export?) of redirects in xls format. This format is severely outdated and doesn't warrant two additional dependencies in Wagtail. If someone has an xls format they can just as easily convert it to csv and import it that way.

## Result

Please consider this proposal, I'm willing to provide PRs to achieve this goal. If accepted the number of Excel file readers/writers would drop from 4 to just one (`openpyxl`).

### PR DESCRIPTION
**Pull request by jaap3**
As mentioned in issue #7374 the amount of packages Wagtail installs to read/write Excel files is suprisingly high. In this PR all Excel file handeling is consolidated to a single library: `openpyxl`. This removes the necessity to install `XlsxWriter`, `tablib`, `xlrd` and `xlwt`, while retaining the ability to export Excel reports and import redirects from (modern) Excel files.

The reduction of dependencies has several benefits:

* Maintainability: reduces the surface of potential breaking updates, API changes etc.
* Security: reduces the amount of code that might have undiscovered security flaws
* Bandwith: reduces the amount of packages that need to be downloaded (in CI, during deploy, dev setup etc.)
* etc.

This PR consists of several logical commits that focus on a single thing. This should hopefully make reviewing these changes a bit easier.

Fixes #7374

### CODE REVIEW

**Comment by jaap3** [comment]
Force pushed some isort fixes (forgot to run it on the last commit). The test failure (test-postgres (3.10 Django main)) is unrelated to this PR.

**Comment by lb-** [comment]
- openyxl is MIT https://foss.heptapod.net/openpyxl/openpyxl/-/blob/branch/3.0/LICENCE.rst
- https://openpyxl.readthedocs.io/en/stable/ - docs link
- it states Excel 2010 format exporting (not sure if that is enough)

Have not researched any further - just adding for reference

**Comment by zerolab** [comment]
Note that is is very much in line with the #6130 goals

File: wagtail/contrib/redirects/utils.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def write_to_file_storage(import_file, input_format):
 
 
 def get_supported_extensions():
-    return ("csv", "tsv", "xls", "xlsx")
+    return "csv", "xlsx"
```
Comment (lb-) [inline wagtail/contrib/redirects/utils.py:20]: does this mean we will no longer support `tsv`  & `xls` exports with this change?

**Review by lb- — COMMENTED** [review]
@jaap3  thanks for looking into this.

Can you please confirm if this reduces / changes any existing support for file formats - looks like this drops support for TSV / XLS.

Which may be ok but will need to be confirmed by core team as this is a breaking change.

Also, can you please rebase on main as this has flagged as having conflicts.

File: wagtail/contrib/redirects/utils.py (hunk 1 above)
Comment (jaap3) [reply wagtail/contrib/redirects/utils.py:20]: imports, exports are unchanged and never supported tsv/xls in the first place.

**Comment by jaap3** [comment]
xls support is definitly dropped. It's a legacy format, the last version of Excel that used xls (instead of xlsx) was Excel 2003. I see no reason to support it, if users have a file in the xls format it can be easily saved as xlsx or csv.

tsv support was there, but only if the file had a .tsv extension. Not sure how common that is, but I've never seen it in the wild. 

If required I could make the csv support a little bit smarter by using a sniffer to detect the flavor of csv (https://docs.python.org/3/library/csv.html#csv.Sniffer), but I'd rather do that in a seperate PR if this lands.

Basically this cuts down the support of import formats to the same set of formats Wagtail is able to export. So data should at least roundtrip.

I'll try to get around to rebasing today or tomorrow. After that I'll be on a holiday for a while. Thanks for looking into this.

**Comment by lb-** [comment]
Thanks for the response - I'll flag this in the next core team meeting.

**Comment by lb-** [comment]
@jaap3 discussed at the core team meeting today and this would be a great addition to Wagtail core.

* Speeds up installs
* Less reliance on multiple packages

Dropping support for XLS is approved, the core team agreed that this format is very old and also has some risk surface for security (macros) that would be nice to not have to think about.

If TSV is still supported with the explicit extension only - that will be fine.

Please ping a comment when ready for a review again (after rebasing).

This will likely not go into Wagtail 4.0 so no time pressure on this week for getting to this.

**Comment by jaap3** [comment]
I've rebased, I won't be able to look into re-adding .tsv support so if that's a blocker it'll have to wait until I get back

**Comment by lb-** [comment]
> I've rebased, I won't be able to look into re-adding .tsv support so if that's a blocker it'll have to wait until I get back

Thanks @jaap3  - please ping when ready.

**Comment by jaap3** [comment]
I've rebased (again) and restored .tsv support, looking forward to getting this merged 😉

**Review by lb- — APPROVED** [review]
This is working great, I have tested a bunch of scenarios locally, export/import (redirects) and cannot see an issue here.

Thank you for pursuing this @jaap3 

I will draft up some release notes and push / rebase to your branch soon.

I will also push the suggested change to update the dependency (this worked locally for me).

@gasman  - I will endeavour to merge this in over the weekend but please take another quick look for anything you see that could be an issue.

File: setup.py [hunk 2]
```diff
@@ -33,8 +33,7 @@
     "Willow>=1.4,<1.5",
     "requests>=2.11.1,<3.0",
     "l18n>=2018.5",
-    "xlsxwriter>=1.2.8,<4.0",
-    "tablib[xls,xlsx]>=0.14.0",
+    "openpyxl>=2.6.4,<4.0",
```
Comment (lb-) [inline setup.py:36]:
```suggestion
    "openpyxl>=3.0.10,<4.0",
```

**Comment by lb-** [comment]
@jaap3  sorry - looks like you have not allowed this PR to be pushed to - that is fine - but please can you apply the updated dependency so the CI can run.

There are no current merge conflicts so no need to rebase again for now.

**Comment by jaap3** [comment]
I've updated the openpyxl pin. Thanks for the suggestion, that makes it easy :). I'm not sure why the option to push was disabled, and cannot find a way to enable it now. I've noticed that this has happened to several of my PRs recently, I'll look into it.

**Comment by jaap3** [comment]
@lb- I've manually gave you write access to the leukeleu/wagtail fork. I do not have the checkbox that's mentioned [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork). I think it has something to do with the fact that this PR comes from an organisations account instead of my user account.

**Comment by lb-** [comment]
Rebased, a few commits squashed, changelog added & pushed.

Will check the CI completes and merge in.

Thanks again @jaap3  - this is an incredible improvement.

File: docs/releases/4.1.md [hunk 3]
```diff
@@ -80,3 +81,8 @@ The following button classes have been removed, none of which were being used wi
 * `hover-no`
 * `unbutton`
 * `yes`
+
+### Changed support for importing Spreadsheet files into Redirects
+
+* `.xls` files will no longer be supported for importing into the contrib Redirects listing, `.xlsx` is still supported.
+* `.tsv` files will now explicitly require the `.tsv` extension to be imported, `.csv` using tab delimiters will no longer work.
```
Comment (jaap3) [inline docs/releases/4.1.md:88]: I believe this was already true before my changes (I didn't verify)

Comment (lb-) [reply docs/releases/4.1.md:88]: thanks.
